### ISSUE TITLE
Heap corruption detected, free list is damaged

### ISSUE THREAD

**Issue by forthrin** [mpv-player/mpv#11378]
### Important Information

Provide following Information:
- mpv version: 4f6d7a
- macOS Version: 13.2.1
- Source of the mpv binary or bundle: mpv-build
- If known which version of mpv introduced the problem: ?
- Possible screenshot or video of visual glitches: N/A

### Reproduction steps

Not intentionally reproducible. Possible duplicate of #10886.

### Expected behavior

No crash

### Actual behavior

```
Exiting... (Quit)
mpv(6838,0x16e7fb000) malloc: Heap corruption detected, free list is damaged at 0x6000032fa880
*** Incorrect guard value: 105553169668688
mpv(6838,0x16e7fb000) malloc: *** set a breakpoint in malloc_error_break to debug
```

### Log file

[mpv.log.gz](https://github.com/mpv-player/mpv/files/10833392/mpv.log.gz)
[mpv.txt.gz](https://github.com/mpv-player/mpv/files/10833442/mpv.txt.gz)

### Sample files

N/A

**Comment by Akemi** [comment]
either this is another case of 'not our problem' or something is 'randomly' corrupting our memory.

**Comment by forthrin** [comment]
With a modern CPU/OS, can C / machine code write to memory outside its own sandbox without being stopped by the kernal? Just wondering, to understand who could be responsible for the overwrite.

I assume `malloc()` mentioned here is the macOS kernel `malloc()`. There are almost no `malloc()` in the mpv source, but FFmpeg have a lot of wrappers/fast replacements. I assume the problem is not in these `malloc()` implementations. The heap just happens to be the target for the overwrite.

**Comment by forthrin** [comment]
This keeps happening quite frequently.

**Comment by low-batt** [comment]
Ignoring special cases like shared memory, the macOS kernel will not let one process write to the memory of another process. The kernel does manage memory for processes, but it is done at the page level, large blocks of memory. An applications obtains pages from the kernel, then it must manage that memory itself. This is where `malloc` comes in, it handles the rapid allocation and deallocation of smaller blocks of memory within the application's process. If the application needs more memory `malloc` will obtain more from the kernel.

Translating this error:
> Heap corruption detected, free list is damaged at 0x6000032fa880
> *** Incorrect guard value: 105553169668688

To do its work `malloc` maintains lists of free memory. Because memory corruption is so insidious there is a need to detect that it has occurred as early as possible. Imagine the application asks `malloc` for 16 bytes. Instead of giving the allocation a 16 byte block `malloc` actually reserves 24 bytes and on either side of the 16 bytes `malloc` writes some sort of known value. When the application gives this memory back to `malloc` the 4 bytes on either side of the 16 byte blocks are checked to make sure they still have the known value. If they don't, then some code somewhere is defective. This is one example of a guard value.

As for what code could be responsible… Exploring into the `mpv` application files shows a number of potential culprits:
```bash
low-batt@gag lib$ pwd
/Applications/mpv.app/Contents/MacOS/lib
low-batt@gag lib$ ls
libarchive.13.dylib	libfreetype.6.dylib	libogg.0.dylib		libssl.3.dylib
libaribb24.0.dylib	libfribidi.0.dylib	libopenjp2.7.dylib	libswresample.4.dylib
libass.9.dylib		libglib-2.0.0.dylib	libopus.0.dylib		libswscale.6.dylib
libavcodec.59.dylib	libgraphite2.3.dylib	libpcre2-8.0.dylib	libuchardet.0.dylib
libavdevice.59.dylib	libharfbuzz.0.dylib	libpng16.16.dylib	libunibreak.5.dylib
libavfilter.8.dylib	libiconv.2.dylib	libpostproc.56.dylib	libvmaf.1.dylib
libavformat.59.dylib	libintl.8.dylib		librubberband.2.dylib	libvorbis.0.dylib
libavutil.57.dylib	liblcms2.2.dylib	libsamplerate.0.dylib	libvpx.7.dylib
libb2.1.dylib		liblua.5.1.dylib	libsodium.23.dylib	libzimg.2.dylib
libcrypto.3.dylib	liblz4.1.dylib		libsoxr.0.dylib		libzmq.5.dylib
libdav1d.6.dylib	liblzma.5.dylib		libsrt.1.5.dylib	libzstd.1.dylib
libfontconfig.1.dylib	libmujs.dylib		libssh.4.dylib
low-batt@gag lib$ 
```

`mpv` is layered on top of a lot of code from other projects.

Xcode's sanitizer tools discussed in [Diagnosing memory, thread, and crash issues early](https://developer.apple.com/documentation/xcode/diagnosing-memory-thread-and-crash-issues-early) may be able to detect a problem. Outside of Xcode `malloc` debugging features can be enabled as discussed in [Enabling the Malloc Debugging Features](https://developer.apple.com/library/archive/documentation/Performance/Conceptual/ManagingMemory/Articles/MallocDebug.html#//apple_ref/doc/uid/20001884-CJBJFIDD).

**Comment by forthrin** [comment]
Thanks for the clarification. Obviously smart to put buffers around each allocated area and check them for overwrites. How can I build a debug version of mpv that maximizes such checking? Can hardware breakpoints be enabled?

**Comment by low-batt** [comment]
I am not a mpv developer. A Mac based mpv developer will have to answer the question about how to build and debug mpv using Xcode. I was unable to figure that out, so I took a different approach. See the issue I entered. It shows how I built a debug version and ran mpv with Guard Malloc as discussed in [Enabling the Malloc Debugging Features](https://developer.apple.com/library/archive/documentation/Performance/Conceptual/ManagingMemory/Articles/MallocDebug.html#//apple_ref/doc/uid/20001884-CJBJFIDD). This was my first time doing this, so a mpv developer may point out errors in my analysis.
